File: demos/cc_fraud/lattigo/debug_helper.go
package fraud_model_lattigo_debug

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"sync"

	"fully_homomorphic_encryption/demos/common/go/pathutils"
	"github.com/tuneinsight/lattigo/v6/core/rlwe"
	"github.com/tuneinsight/lattigo/v6/schemes/ckks"
)

var (
	refData     map[string]map[string][]float64
	loadOnce    sync.Once
	loadErr     error
	debugRowIdx int = -1
)

func loadReferenceData() {
	path := pathutils.ResolvePath("fully_homomorphic_encryption/demos/cc_fraud/debug/debug_reference.json")
	var file *os.File
	file, loadErr = os.Open(path)
	if loadErr != nil {
		return
	}
	defer file.Close()

	decoder := json.NewDecoder(file)
	loadErr = decoder.Decode(&refData)
}

func getRowIdx() int {
	if debugRowIdx != -1 {
		return debugRowIdx
	}

	rowStr := os.Getenv("HEIR_DEBUG_ROW_IDX")
	if rowStr == "" {
		// Fallback to reading from a temp file or just default to 0
		// In Go, it might be harder to pass flags to the helper, so env var is best.
		rowStr = "0"
	}

	idx, err := strconv.Atoi(rowStr)
	if err != nil {
		fmt.Printf("  [DEBUG] Invalid HEIR_DEBUG_ROW_IDX '%s', defaulting to 0\n", rowStr)
		debugRowIdx = 0
	} else {
		debugRowIdx = idx
	}
	return debugRowIdx
}

var allowedSteps = map[string]bool{
	"input":          true,
	"layer1_matmul":  true,
	"layer1_bias":    true,
	"layer1_sigmoid": true,
	"layer2_matmul":  true,
	"layer2_bias":    true,
	"layer2_sigmoid": true,
	"layer3_matmul":  true,
	"layer3_bias":    true,
}

func __heir_debug(evaluator *ckks.Evaluator, param ckks.Parameters, encoder *ckks.Encoder, decryptor *rlwe.Decryptor, ctObj any, debugAttrMap map[string]string) {
	loadOnce.Do(loadReferenceData)

	var ct *rlwe.Ciphertext
	switch v := ctObj.(type) {
	case *rlwe.Ciphertext:
		ct = v
	case []*rlwe.Ciphertext:
		if len(v) == 0 {
			fmt.Println("  [DEBUG] Empty ciphertext slice")
			return
		}
		ct = v[0]
		if ct == nil {
			fmt.Println("  [DEBUG] First ciphertext element is nil")
			return
		}
	default:
		panic(fmt.Sprintf("unexpected type %T", ctObj))
	}

	stepName := debugAttrMap["debug.name"]
	if !allowedSteps[stepName] {
		return
	}

	rowIdx := getRowIdx()
	rowKey := fmt.Sprintf("row_%d", rowIdx)

	// Determine size to decode
	messageSizeStr, ok := debugAttrMap["message.size"]
	var messageSize int
	var err error
	if !ok || messageSizeStr == "" {
		messageSize = 1
	} else {
		messageSize, err = strconv.Atoi(messageSizeStr)
		if err != nil {
			messageSize = 1
		}
	}

	// Decrypt and decode
	pt := decryptor.DecryptNew(ct)
	values := make([]float64, param.MaxSlots())
	encoder.Decode(pt, values)

	// Slice to actual size
	fheVals := values[:messageSize]

	// Print FHE values
	printSize := 5
	if len(fheVals) < printSize {
		printSize = len(fheVals)
	}
	fmt.Printf("[DEBUG] Step: %s (row_%d)\n", stepName, rowIdx)
	fmt.Printf("  FHE Decrypted (first min(5, size)): %v (size: %d)\n", fheVals[:printSize], len(fheVals))
	fmt.Printf("  Scale: 2^%3.3f\n", ct.Scale.Log2())

	if loadErr != nil {
		fmt.Printf("  [WARNING] Failed to load reference data: %v\n", loadErr)
		return
	}

	// Compare with reference
	if rowData, ok := refData[rowKey]; ok {
		if refVals, ok := rowData[stepName]; ok {
			refPrintSize := 5
			if len(refVals) < refPrintSize {
				refPrintSize = len(refVals)
			}
			fmt.Printf("  Expected Ref  (first min(5, size)): %v\n", refVals[:refPrintSize])

			// Slice FHE values to match reference size
			compareFheVals := fheVals
			if len(fheVals) > len(refVals) {
				compareFheVals = fheVals[:len(refVals)]
			}

			// Calculate max absolute error
			maxAbsErr := 0.0
			for i := 0; i < len(compareFheVals); i++ {
				err := math.Abs(compareFheVals[i] - refVals[i])
				if err > maxAbsErr {
					maxAbsErr = err
				}
			}
			fmt.Printf("  Max Abs Error: %e\n", maxAbsErr)
			if maxAbsErr > 0.0 {
				fmt.Printf("  Precision Lost: 2^%3.3f bits\n", math.Log2(maxAbsErr))
			} else {
				fmt.Println("  Precision Lost: 0 bits (exact)")
			}

			// Sorted Check
			if len(compareFheVals) == len(refVals) {
				sortedFhe := make([]float64, len(compareFheVals))
				copy(sortedFhe, compareFheVals)
				sortedRef := make([]float64, len(refVals))
				copy(sortedRef, refVals)

				sort.Float64s(sortedFhe)
				sort.Float64s(sortedRef)

				maxSortedErr := 0.0
				for i := 0; i < len(sortedFhe); i++ {
					err := math.Abs(sortedFhe[i] - sortedRef[i])
					if err > maxSortedErr {
						maxSortedErr = err
					}
				}
				fmt.Printf("  [Sorted Check] Max Abs Error: %e\n", maxSortedErr)
				if maxSortedErr > 0.0 {
					fmt.Printf("  [Sorted Check] Precision Lost: 2^%3.3f bits\n", math.Log2(maxSortedErr))
				} else {
					fmt.Println("  [Sorted Check] Precision Lost: 0 bits (exact)")
				}
			} else {
				fmt.Printf("  [Sorted Check] Skip (size mismatch: FHE %d vs Ref %d)\n", len(compareFheVals), len(refVals))
			}

		} else {
			fmt.Printf("  [WARNING] No reference data found for step '%s'\n", stepName)
		}
	} else {
		fmt.Printf("  [WARNING] No reference data found for row '%s'\n", rowKey)
	}
}

File: demos/cc_fraud/openfhe/debug_helper.cpp
#include "demos/cc_fraud/openfhe/debug_helper.h"

#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <map>
#include <string>
#include <vector>

#include "demos/cc_fraud/openfhe/debug_reference.h"
#include "src/pke/include/ciphertext.h"
#include "src/pke/include/cryptocontext.h"
#include "src/pke/include/encoding/plaintext.h"

using PlaintextT = lbcrypto::Plaintext;

void __heir_debug(CryptoContextT cc, PrivateKeyT sk, CiphertextT ct,
                  const std::map<std::string, std::string>& debugAttrMap) {
  // Get op name
  std::string op_name = "unknown";
  if (debugAttrMap.find("debug.name") != debugAttrMap.end()) {
    op_name = debugAttrMap.at("debug.name");
  } else if (debugAttrMap.find("asm.op_name") != debugAttrMap.end()) {
    op_name = debugAttrMap.at("asm.op_name");
  }

  // Get current row index from environment
  const char* row_idx_str = std::getenv("HEIR_DEBUG_ROW_IDX");
  std::string row_key = "row_0";
  if (row_idx_str) {
    row_key = "row_" + std::string(row_idx_str);
  }

  std::cout << "\n[DEBUG] Step: " << op_name << " (" << row_key << ")"
            << std::endl;

  // Decrypt
  PlaintextT ptxt;
  cc->Decrypt(sk, ct, &ptxt);

  // We need to know how many elements to print.
  // We can get it from the reference data if available, or default to a small
  // number.
  size_t print_size = 5;
  bool has_ref = false;
  std::vector<float> ref_vals;

  if (debug_reference.find(row_key) != debug_reference.end()) {
    const auto& row_ref = debug_reference.at(row_key);
    if (row_ref.find(op_name) != row_ref.end()) {
      ref_vals = row_ref.at(op_name);
      print_size = ref_vals.size();
      has_ref = true;
    }
  }

  // Set length of plaintext to decode
  // If we don't have ref, we might decode too many slots.
  // In CKKS, we usually only care about the active slots.
  // If we don't have ref, we can try to get it from message.size if present.
  if (!has_ref && debugAttrMap.find("message.size") != debugAttrMap.end()) {
    print_size = std::stoul(debugAttrMap.at("message.size"));
  }

  ptxt->SetLength(print_size);

  // CKKS decrypts to CKKSPackedEncoding which can be cast to vector of complex
  // or double. In OpenFHE, CKKSPackedEncoding has GetRealHSV() to get real
  // values.
  auto fhe_vals = ptxt->GetRealPackedValue();

  // Print FHE values
  std::cout << "  FHE Decrypted (first min(5, size)): [";
  for (size_t i = 0; i < std::min(fhe_vals.size(), (size_t)5); ++i) {
    std::cout << fhe_vals[i]
              << (i == std::min(fhe_vals.size(), (size_t)5) - 1 ? "" : ", ");
  }
  if (fhe_vals.size() > 5) std::cout << ", ...";
  std::cout << "] (size: " << fhe_vals.size() << ")" << std::endl;

  // Print Scale
  double scale = ct->GetScalingFactor();
  std::cout << "  Scale: 2^" << std::log2(scale) << std::endl;

  // Compare with reference
  if (has_ref) {
    std::cout << "  Expected Ref  (first min(5, size)): [";
    for (size_t i = 0; i < std::min(ref_vals.size(), (size_t)5); ++i) {
      std::cout << ref_vals[i]
                << (i == std::min(ref_vals.size(), (size_t)5) - 1 ? "" : ", ");
    }
    if (ref_vals.size() > 5) std::cout << ", ...";
    std::cout << "]" << std::endl;

    // Calculate precision loss
    double max_abs_err = 0.0;
    for (size_t i = 0; i < std::min(fhe_vals.size(), ref_vals.size()); ++i) {
      double err = std::abs(fhe_vals[i] - ref_vals[i]);
      if (err > max_abs_err) {
        max_abs_err = err;
      }
    }
    std::cout << "  Max Abs Error: " << max_abs_err << std::endl;
    if (max_abs_err > 0.0) {
      std::cout << "  Precision Lost: 2^" << std::log2(max_abs_err) << " bits"
                << std::endl;
    } else {
      std::cout << "  Precision Lost: 0 bits (exact)" << std::endl;
    }

    // Sorted comparison to check if it is just a permutation
    if (fhe_vals.size() == ref_vals.size()) {
      std::vector<double> sorted_fhe = fhe_vals;
      std::vector<float> sorted_ref = ref_vals;
      std::sort(sorted_fhe.begin(), sorted_fhe.end());
      std::sort(sorted_ref.begin(), sorted_ref.end());

      double max_sorted_err = 0.0;
      for (size_t i = 0; i < sorted_fhe.size(); ++i) {
        double err = std::abs(sorted_fhe[i] - sorted_ref[i]);
        if (err > max_sorted_err) {
          max_sorted_err = err;
        }
      }
      std::cout << "  [Sorted Check] Max Abs Error: " << max_sorted_err
                << std::endl;
      if (max_sorted_err > 0.0) {
        std::cout << "  [Sorted Check] Precision Lost: 2^"
                  << std::log2(max_sorted_err) << " bits" << std::endl;
      } else {
        std::cout << "  [Sorted Check] Precision Lost: 0 bits (exact)"
                  << std::endl;
      }
    } else {
      std::cout << "  [Sorted Check] Skip (size mismatch: FHE "
                << fhe_vals.size() << " vs Ref " << ref_vals.size() << ")"
                << std::endl;
    }
  } else {
    std::cout << "  [WARNING] No reference data found for this step."
              << std::endl;
  }
}

void __heir_debug(CryptoContextT cc, PrivateKeyT sk,
                  std::vector<CiphertextT> cts,
                  const std::map<std::string, std::string>& debugAttrMap) {
  if (!cts.empty()) {
    __heir_debug(cc, sk, cts[0], debugAttrMap);
  }
}
